Include/internal/pycore_object.h

```diff
@@ -220,7 +224,46 @@ _Py_DECREF_SPECIALIZED(PyObject *op, const destructor destruct)
 static inline void
 _Py_DECREF_NO_DEALLOC(PyObject *op)
 {
-    Py_DECREF(op);
+    if (_Py_IsImmortal(op)) {
+        return;
+    }
+
+    if (_Py_IsOwnedByCurrentThread(op)) {
+        uint32_t refcount = _Py_atomic_load_uint32_relaxed(&op->ob_ref_local);
+        assert(refcount != 0);
+        refcount--;
+        _Py_atomic_store_uint32_relaxed(&op->ob_ref_local, refcount);
+        if (refcount == 0) {
+            // Assume that local + share >= 1
+            _Py_MergeZeroLocalRefcount(op);
+        }
+    }
+    else {
+        Py_ssize_t refcount = _Py_atomic_load_ssize_relaxed(&op->ob_ref_shared);
+        Py_ssize_t new_shared;
+        // Shared refcount can be zero but we should consider local refcount.
+        int should_queue = (refcount == 0 || refcount == _Py_REF_MAYBE_WEAKREF);
```

Comment:
@colesbury Out of curiosity, Don't we have to consider that shared_refcount can be a negative value at this moment due to imbalance refcounting?

Same question for the 
https://github.com/python/cpython/blob/244567398370bfa62a332676762bb1db395c02fc/Objects/object.c#L319

Reply:
I'm not sure I fully understand your question. Yes, `shared_refcount` may be negative. `should_queue` is always false if `shared` is negative because:

1) we only queue objects once
2) we queue them the first time the shared refcount becomes negative

So if it's already negative then we must have already queued it.

Reply:
> So if it's already negative then we must have already queued it.

Make sense, thank you for explain.